grpc/interceptors.go

```diff
@@ -13,21 +16,59 @@ import (
 	"github.com/letsencrypt/boulder/features"
 )
 
+const (
+	returnOverhead         = 20 * time.Millisecond
+	meaningfulWorkOverhead = 100 * time.Millisecond
+	clientRequestTimeKey   = "client-request-time"
+	serverLatencyKey       = "server-latency"
+)
+
 // serverInterceptor is a gRPC interceptor that adds Prometheus
 // metrics to requests handled by a gRPC server, and wraps Boulder-specific
 // errors for transmission in a grpc/metadata trailer (see bcodes.go).
 type serverInterceptor struct {
 	serverMetrics *grpc_prometheus.ServerMetrics
+	rpcLag        prometheus.Histogram
+	clk           clock.Clock
 }
 
-const returnOverhead = 20 * time.Millisecond
-const meaningfulWorkOverhead = 100 * time.Millisecond
+func newServerInterceptor(metrics serverMetrics, clk clock.Clock) serverInterceptor {
+	return serverInterceptor{
+		serverMetrics: metrics.GRPCMetrics,
+		rpcLag:        metrics.RPCLag,
+		clk:           clk,
+	}
+}
 
 func (si *serverInterceptor) intercept(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
 	if info == nil {
 		return nil, berrors.InternalServerError("passed nil *grpc.UnaryServerInfo")
 	}
 
+	// Extract the grpc metadata from the context
+	md, ok := metadata.FromContext(ctx)
+	if !ok {
+		return nil, berrors.InternalServerError("passed context with no grpc metadata")
+	}
+	// Ensure only one value is present
+	if len(md[clientRequestTimeKey]) != 1 {
```

Comment:
This doesn't meet the deployability guidelines. We have to tolerate this field being absent. So maybe wrap the whole "observe time" section into a function, and call it only if the length is >0?